source/darwin/nw_socket.c

```diff
@@ -0,0 +1,2075 @@
+}
+
+static int s_unlock_socket_state(struct nw_socket *nw_socket) {
+    return aws_mutex_unlock(&nw_socket->synced_state.lock);
+}
+
+static bool s_validate_event_loop(struct aws_event_loop *event_loop) {
+    return event_loop && event_loop->vtable && event_loop->impl_data;
+}
+
+static void s_set_event_loop(struct aws_socket *aws_socket, struct aws_event_loop *event_loop) {
+    aws_socket->event_loop = event_loop;
+    struct nw_socket *nw_socket = aws_socket->impl;
+    // Never re-assign an event loop
+    AWS_ASSERT(nw_socket->event_loop == NULL);
+    nw_socket->event_loop = event_loop;
+
+    AWS_LOGF_DEBUG(AWS_LS_IO_SOCKET, "id=%p: s_set_event_loop: socket acquire event loop group.", (void *)nw_socket);
+    aws_event_loop_group_acquire(get_base_event_loop_group(event_loop));
+}
+
+static void s_release_event_loop(struct nw_socket *nw_socket) {
+    if (nw_socket->event_loop == NULL) {
+        AWS_LOGF_DEBUG(AWS_LS_IO_SOCKET, "id=%p: s_release_event_loop: socket has not event loop.", (void *)nw_socket);
+        return;
+    }
+    aws_event_loop_group_release(get_base_event_loop_group(nw_socket->event_loop));
+    AWS_LOGF_DEBUG(
+        AWS_LS_IO_SOCKET, "id=%p: s_release_event_loop: socket release event loop group.", (void *)nw_socket);
+    nw_socket->event_loop = NULL;
+}
+
+static inline void s_set_socket_state(struct nw_socket *nw_socket, struct aws_socket *socket, int state) {
+    s_lock_socket_state(nw_socket);
+    // The state can only go increasing, except for LISTENING and STOPPED. They can switch between each other.
+    if (nw_socket->synced_state.state < state || (state == LISTENING && nw_socket->synced_state.state == STOPPED)) {
+        nw_socket->synced_state.state = state;
+        if (socket) {
+            socket->state = state;
+        }
+    }
+    s_unlock_socket_state(nw_socket);
+}
+
+static int s_setup_socket_params(struct nw_socket *nw_socket, const struct aws_socket_options *options) {
+    if (options->type == AWS_SOCKET_STREAM) {
+        /* if TCP, setup all the tcp options */
+        if (options->domain == AWS_SOCKET_IPV4 || options->domain == AWS_SOCKET_IPV6) {
+            // DEBUG WIP NW_PARAMETERS_DISABLE_PROTOCOL will need to be changed to use MTLS With SecItem
+            nw_socket->socket_options_to_params =
+                nw_parameters_create_secure_tcp(NW_PARAMETERS_DISABLE_PROTOCOL, ^(nw_protocol_options_t nw_options) {
+                  if (options->connect_timeout_ms) {
+                      /* this value gets set in seconds. */
+                      nw_tcp_options_set_connection_timeout(
+                          nw_options, options->connect_timeout_ms / AWS_TIMESTAMP_MILLIS);
+                  }
+
+                  // Only change default keepalive values if keepalive is true and both interval and timeout
+                  // are not zero.
+                  if (options->keepalive && options->keep_alive_interval_sec != 0 &&
```

Comment:
Have we tested (manually with packet trace) that TCP keep alive works as expected?  There were subtle differences between windows and posix and we made some questionable API decisions there.

Reply:
Tested on posix and nw_socket, the socket will send the keep alive packet every `keep_alive_interval_sec` seconds, and expecting ACK within `keep_alive_timeout_sec` seconds.